src/diagonal.works/b6/api/functions/graph.go

```diff
@@ -247,42 +249,59 @@ func WeightsFromOptions(options b6.UntypedCollection, w b6.World) (graph.Weights
 	if err != nil {
 		return nil, err
 	}
+	return WeightsFromTags(opts, w)
+}
 
+func WeightsFromTags(opts b6.Tags, w b6.World) (graph.Weights, error) {
 	var weights graph.Weights
 
-	if opts.Get("elevation").IsValid() {
-		elevation := graph.ElevationWeights{}
-		if upHill := opts.Get("uphill"); upHill.IsValid() && upHill.Value.String() == "hard" {
-			elevation.UpHillHard = true
-		}
-		if downHill := opts.Get("downhill"); downHill.IsValid() && downHill.Value.String() == "hard" {
-			elevation.DownHillHard = true
-		}
-		elevation.W = w
-		weights = elevation
-	} else {
+	switch m := opts.Get("mode").Value.String(); m {
+	case "", "walk":
 		walking := graph.WalkingTimeWeights{
 			Speed: graph.WalkingMetersPerSecond,
 		}
-
-		if speed := opts.Get("walking speed"); speed.IsValid() {
+		if speed := opts.Get("walk:speed"); speed.IsValid() {
 			if f, err := strconv.ParseFloat(speed.Value.String(), 64); err == nil {
 				walking.Speed = f
+			} else {
+				return nil, fmt.Errorf("expected a float string for walk:speed, found %q", speed.Value.String())
 			}
 		}
 		weights = walking
-	}
-
-	switch m := opts.Get("mode").Value.String(); m {
-	case "", "walk":
+		if opts.Get("elevation").IsValid() {
+			elevation := graph.ElevationWeights{
+				NaismithPenalty: graph.NaismithPenalty,
+				Weights:         weights,
+			}
+			if naismith := opts.Get("elevation:naismith"); naismith.IsValid() {
+				if f, err := strconv.ParseFloat(naismith.Value.String(), 64); err == nil {
+					elevation.NaismithPenalty = f
+				} else {
+					return nil, fmt.Errorf("expected a float string for elevation:naismith, found %q", naismith.Value.String())
+				}
+			}
+			if upHill := opts.Get("elevation:uphill"); upHill.IsValid() && upHill.Value.String() == "hard" {
+				elevation.UpHillHard = true
+			}
+			if downHill := opts.Get("elevation:downhill"); downHill.IsValid() && downHill.Value.String() == "hard" {
+				elevation.DownHillHard = true
+			}
+			elevation.W = w
+			weights = elevation
+		}
 	case "transit":
+		opts.ModifyOrAddTag(b6.Tag{Key: "mode", Value: b6.String("walk")})
+		walking, err := WeightsFromTags(opts, w)
+		if err != nil {
+			return nil, err
+		}
 		if p := opts.Get("peak"); p.Value.String() == "no" {
-			weights = graph.TransitTimeWeights{PeakTraffic: false, Weights: weights}
+			weights = graph.TransitTimeWeights{PeakTraffic: false, Weights: walking}
 		} else {
-			weights = graph.TransitTimeWeights{PeakTraffic: true, Weights: weights}
```

Comment:
so the reason i did it the other way is b/c i wanted to be able to combine elevation w transit / which we want (i think / or at least i want)